### PR TITLE
basic InductorBenchmarker

### PR DESCRIPTION
**Pull request by nmacchioni**
This PR adds the most basic custom benchmarker (i.e. a benchmarker that is not provided by Triton), which we call `InductorBenchmarker`. This new benchmarker is very basic in principal, and very closely follows Triton's `do_bench` implementation with slight changes such as flushing the exact L2 cache size (Triton defaults to 256mb), using a buffer zero for warmup (Triton uses the benchmarked kernel itself, I found that buffer zeroes are more consistent),  and returning the min runtime (Triton can return min, among other things, currently Inductor picks median).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #144507
* #144505
* #144501
* #144353
* #133287
* #144365
* #133121
* __->__ #133058
* #144315

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

### CODE REVIEW

File: torch/_inductor/runtime/benchmarking.py [hunk 1]
```diff
@@ -195,11 +198,189 @@ def benchmark_gpu(self: Self, _callable: Callable[[], Any], **kwargs: Any) -> fl
         this is the first requested quantile. Else, if `kwargs["return_mode"]` is specified,
         this is the requested return mode. Otherwise, this is the median.
         """
+
+        # this method may be used as a fallback if certain benchmarking features are disabled,
+        # in which case those requests may contain kwargs that are not specific to Triton's
+        # `do_bench_gpu`. as such, we may need to prune out kwargs that do not exists in
+        # `do_bench_gpu`'s signature.
+        do_bench_params = inspect.signature(self.triton_do_bench).parameters
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:206]: should we cache this ?

File: torch/_inductor/runtime/benchmarking.py [hunk 2]
```diff
@@ -195,11 +198,189 @@ def benchmark_gpu(self: Self, _callable: Callable[[], Any], **kwargs: Any) -> fl
...
+        for kwarg in list(kwargs.keys()):
+            if kwarg not in do_bench_params:
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:208]: warn maybe ?

File: torch/_inductor/config.py [hunk 3]
```diff
@@ -1335,6 +1335,53 @@ class trace:
     log_autotuning_results: bool = False
 
 
+def get_feature_local_version(feature_name: str) -> Optional[int]:
```
Comment (eellison) [inline torch/_inductor/config.py:1338]: How does this relate to JK ?

File: torch/_inductor/runtime/benchmarking.py [hunk 4]
```diff
@@ -195,11 +198,189 @@ def benchmark_gpu(self: Self, _callable: Callable[[], Any], **kwargs: Any) -> fl
...
+    feature_config = getattr(benchmarking_config, feature_name)
+    if feature_config.env_val is not None:
+        if feature_config.env_val == "1":
+            return True
+        elif feature_config.env_val == "0":
+            return False
+    if not is_fbcode():
+        return feature_config.oss_default
+    if feature_config.local_version is not None:
+        return (
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:231]: Feel like this could be standardized with some of our other tri state/ jk checking.. although I have been less involved with that. cc @masnesral  @oulgen

File: torch/_inductor/runtime/benchmarking.py [hunk 5]
```diff
@@ -195,11 +198,189 @@ def benchmark_gpu(self: Self, _callable: Callable[[], Any], **kwargs: Any) -> fl
...
+                for start_event, end_event in event_pairs
+            ]
+        )
+
+    @maybe_fallback
+    @maybe_time
+    @count
+    def benchmark_gpu(
+        self: Self,
+        _callable: Callable[[], Any],
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:308]: Just for posterity - how were these chosen and how do they differ from triton's function ?

**Review by eellison — APPROVED** [review]
Looks great ! Note for other reviewers: this isnt turned on by default yet..

also cc @masnesral to review

File: torch/_inductor/runtime/benchmarking.py (hunk 4 above)
Comment (c00w) [reply torch/_inductor/runtime/benchmarking.py:231]: Yeah - FYI - you can already do allmost of this via Config objects in the config
https://github.com/pytorch/pytorch/blob/1986b46d6349f3692c1a25ba9a3f1fcb8006751c/torch/utils/_config_module.py#L22

i.e. this is almost equivalent to benchmarking_config.feature_x = Config(env_value_default="env_name, default=oss_default, justknob-"pytorch/benchmarking:feature_name")

Note that today we don't support the version pattern (and I'm working on making JK commit time aware, so I can just kill it totally and make it a native feature of JK).

If you feel strongly about wanting to use versions in JK with config objects, I can stack a PR to add support..

File: torch/_inductor/runtime/benchmarking.py (hunk 1 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/benchmarking.py]: I simplified this a lot, we no longer have fallbacks (just a single global toggle) so we don't have to worry about this case

File: torch/_inductor/runtime/benchmarking.py (hunk 2 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/benchmarking.py]: same as above, I removed the fallbacks so this is thankfully no longer required

File: torch/_inductor/config.py (hunk 3 above)
Comment (nmacchioni) [reply torch/_inductor/config.py]: switched to global toggle

File: torch/_inductor/runtime/benchmarking.py (hunk 4 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/benchmarking.py]: I moved away from using a JK for now, just a global toggle, if we feel a JK is useful later I can implement that in a separate PR

File: torch/_inductor/runtime/benchmarking.py (hunk 5 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/benchmarking.py:202]: So the params are very similar, mostly we change the warmup/benchmark specification from # of ms to # of iterations. We also add memory warmup iters, which is new from Triton's do_bench, to control the number of memory warmup repetitions

File: torch/_inductor/runtime/benchmarking.py [hunk 6]
```diff
@@ -157,11 +159,129 @@ def benchmark_gpu(self: Self, _callable: Callable[[], Any], **kwargs: Any) -> fl
...
+        of `memory_warmup_iters` and `benchmark_iters`, along with the estimated
+        runtime of `_callable` and various other factors, and we then shrink
+        `benchmark_iters` to fit in the alloted maximum duration.
+        - **kwargs: Additional kwargs that may be passed to the fallback.
+
+        Returns:
+        - The minimum runtime of `_callable`, in milliseconds.
+        """
+        # we don't want any outside errors propagating into benchmarking
+        torch.cuda.synchronize()
```
Comment (etaf) [inline torch/_inductor/runtime/benchmarking.py:237]: Can you use common device interface instead of hard cuda here? This will break benchmark on XPU.

**Comment by etaf** [comment]
Hi, I've learned the series of your PRs, and suggested to use common device interface instead of use hard code torch.cuda.xxx. Because origin `benchmark_gpu` supports both CUDA and XPU, but your change will break XPU.

File: benchmarks/dynamo/pr_time_benchmarks/expected_results.csv [hunk 7]
```diff
@@ -30,7 +30,7 @@ basic_modules_ListOfLinears_inductor_gpu_force_shape_pad,compile_time_instructio
 
 
 
-basic_modules_ListOfLinears_inductor_gpu,compile_time_instruction_count,17260000000,0.2
+basic_modules_ListOfLinears_inductor_gpu,compile_time_instruction_count,10885050825,0.2
```
Comment (eellison) [inline benchmarks/dynamo/pr_time_benchmarks/expected_results.csv:33]: nice !

File: torch/_inductor/config.py [hunk 8]
```diff
@@ -316,6 +316,12 @@ def bundle_triton_into_fx_graph_cache_default() -> Optional[bool]:
 # default value is InfiniBand
 inter_node_bw = 25
 
+# use Inductor's experimental benchmarker (runtime/benchmarking.py)
+# to benchmark kernels during autotuning, otherwise fall back to
+# Triton's `do_bench`. the experimental benchmarker may produce
+# results that are not consistent with `do_bench`'s results
+use_experimental_benchmarker = True
```
Comment (eellison) [inline torch/_inductor/config.py:323]: nit - use this [Config](https://github.com/pytorch/pytorch/blob/3908be676c7406da52824c83b804ae45e1daff7a/torch/utils/_config_module.py#L47) class that will JK if we need it.

File: torch/_inductor/runtime/benchmarking.py (hunk 6 above)
Comment (eellison) [reply torch/_inductor/runtime/benchmarking.py:237]: let's address this before landing - triton do_bench uses device interface so this would be a regression

**Review by eellison — COMMENTED** [review]
looks good, just needs to update to generic device synchronize

File: benchmarks/dynamo/pr_time_benchmarks/expected_results.csv (hunk 7 above)
Comment (nmacchioni) [reply benchmarks/dynamo/pr_time_benchmarks/expected_results.csv:33]: what does this mean btw?

Comment (eellison) [reply benchmarks/dynamo/pr_time_benchmarks/expected_results.csv:33]: lower compile time instruction count

File: torch/_inductor/runtime/benchmarking.py [hunk 9]
```diff
@@ -7,9 +8,11 @@
 
 import torch
 from torch._dynamo.utils import counters, dynamo_timed
+from torch._inductor.config import use_experimental_benchmarker
 
 
 logger = torch._logging.getArtifactLogger(__name__, "benchmarking")
+use_experimental_benchmarker = use_experimental_benchmarker and not torch.xpu._is_compiled
```
Comment (nmacchioni) [inline torch/_inductor/runtime/benchmarking.py:15]: @etaf we're going to force disable the experimental benchmarker on xpu for now; it's not so straightforward to figure out if we're on xpu when `benchmark_gpu` is called since the tensors are already wrapped away

Comment (etaf) [reply torch/_inductor/runtime/benchmarking.py]: @nmacchioni  thanks for your information, but I suggest using the common device interface in `torch/_dynamo/device_interface.py` so that the code is more generic and will easy to enable for other GPU.

Comment (nmacchioni) [reply torch/_inductor/runtime/benchmarking.py]: @etaf we can't use `torch/_dynamo/device_interface.py` since that requires `device_type`, we don't know `device_type` in `benchmark_gpu` since `_callable` is essentially a black box kernel. We'd either have to migrate all call sites to `benchmark` or modify them to pass `device_type` to `benchmark_gpu`, neither of which I am interested in doing at the moment.

Comment (etaf) [reply torch/_inductor/runtime/benchmarking.py]: @nmacchioni  Maybe use `get_gpu_type()` in `torch/_inductor/utils.py` ?

Comment (nmacchioni) [reply torch/_inductor/runtime/benchmarking.py]: @etaf Under the hood is this not more or less `torch.xpu._is_compiled()`?

Regardless, I don't feel that it is particularly safe to enable this outside of Nvidia/AMD hardware at the moment; there is no guarantee that this specialized benchmarking module will work as expected on other hardware, and I certainly haven't done the required performance for any other hardware.

Comment (etaf) [reply torch/_inductor/runtime/benchmarking.py]: @nmacchioni , agree with you that it only works on CUDA at the moment, what I meant was that using a common device interface would make the code more generic and extensible, not that it should work on other GPUs now. If this is going to increase your work, we can wait to do the generalization work later when other GPUs need it ^_^.

Comment (nmacchioni) [reply torch/_inductor/runtime/benchmarking.py]: @etaf Sounds good! We can always work on an XPU focused benchmarked in the future, should only require some slight algorithmic tweaks

**Comment by laithsakka** [comment]
huge win!
